python-auto-metadata-collection/ai4hf_passport_base.py

```diff
@@ -0,0 +1,420 @@
+        response.raise_for_status()
+        return response.json()
+    
+    def submit_results_to_ai4hf_passport(self, 
+                                        model: Any, 
+                                        learning_stages: list[LearningStage], 
+                                        evaluation_measures: list[EvaluationMeasure],
+                                        model_info: Model):
+        """
+        Submit results of the ML model to the AI4HF Passport Server.
+
+        :param model: Model class that specific to implemented library.
+        :param learning_stages: The list of learning stages.
+        :param evaluation_measures: The list of evaluation measures.
+        :param model_info: Model class for model related fields.
+
+        """
+        print('Sending given informations into AI4HF Passport server....')
+        learning_process = self.extract_and_submit_learning_process(model)
+        print(f'Learning process created: {learning_process}')
+
+        for learning_stage in learning_stages:
+            learning_stage.learningProcessId = learning_process.learningProcessId
+            learning_stage_response = self.submit_learning_stage(learning_stage)
+            print(f'Learning stage created: {learning_stage_response}')
+        
+        user_id = jwt.decode(self.token, options={"verify_signature": False})['sub']
+        model_info.learningProcessId = learning_process.learningProcessId
+        model_info.studyId = self.study_id
+        model_info.owner = self.organization_id
+        model_info.createdBy = user_id
+        model_info.lastUpdatedBy = user_id
+        created_model: Model = self.extract_and_submit_model(model, model_info)
+        print(f'Model created: {created_model}')
+
+        for evaluation_measure in evaluation_measures:
+            evaluation_measure.modelId = created_model.modelId
+            evaluation_measure_response = self.submit_evaluation_measure(evaluation_measure)
+            print(f'Learning stage created: {evaluation_measure_response}')
+        
+        self.extract_and_submit_parameters(model, created_model.modelId)
+        print('Given informations are sent into AI4HF Passport server!')
+        pass
+
+    def submit_parameter(self, parameter: Parameter) -> Parameter:
+        """
+        Submit parameter to the AI4HF Passport Server.
+
+        :param parameter: Parameter object to be sent.
+
+        :return parameter: Created parameter object from the server.
+        """
+        url = f"{self.passport_server_url}/parameter?studyId={self.study_id}"
+        headers = {"Authorization": f"Bearer {self.token}", "Content-Type": "application/json"}
+        payload = {"studyId": parameter.studyId, "name": parameter.name, 
+                   "dataType": parameter.dataType, "description": parameter.description}
+        response = requests.post(url, json=payload, headers=headers)
+
+        if response.status_code == 401:  # Token expired, refresh and retry
```

Comment:
I think this part is duplicated (headers and response.post are also duplicated), let's create a method to handle this post request in addition to the token refresh.